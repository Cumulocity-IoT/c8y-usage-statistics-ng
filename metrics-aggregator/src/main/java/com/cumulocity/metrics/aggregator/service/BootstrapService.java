package com.cumulocity.metrics.aggregator.service;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.event.EventListener;
import org.springframework.stereotype.Service;
import com.cumulocity.microservice.subscription.model.MicroserviceSubscriptionAddedEvent;






/**
 * This is a service to fetch subtenants etc
 * 
 * @author Marco Stoffel
 *
 */

@Service
public class BootstrapService {

    @Autowired
    DeviceMetricsAggregationService deviceMetricsAggregationService;
 
    @Autowired
    StartupUtils utils;

    private static final Logger log = LoggerFactory.getLogger(BootstrapService.class);

    @EventListener
	public void initialize(MicroserviceSubscriptionAddedEvent event) {
		log.info("Tenant Sub: " + event.toString());
        utils.setCurrentTenant(event.getCredentials().getTenant());
        utils.getTenants();
        deviceMetricsAggregationService.createDailyDeviceStatistics();
	}

}
