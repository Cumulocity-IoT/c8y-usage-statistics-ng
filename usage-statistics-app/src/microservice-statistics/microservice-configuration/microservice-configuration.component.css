.prod-cat-cont {
    display: flex;
    justify-content: space-between;
    align-items: center;
}

.prod-cat-cont-item {
    display: flex;
    gap: 8px;
    align-items: center;
}

.prod-cancel {
    transform: translateY(8%);
}

.pointer{
    cursor: pointer;
}
.hide-item{
    visibility: hidden;
}

.alert-info-custom{
    width: 90%;
    margin: auto;
    margin-bottom: 20px;
}

.sticky-card-bottom{
    position: sticky;
    bottom: -20.27px;
    background-color: white;
    z-index: 10;
}

@media only screen and (min-width: 1200px ) {
    .fixed{
        position: fixed;
    }    
}

@media only screen and (max-width: 600px) {
    .mob{
        display: flex;
        gap: 10px
    }
    .btn{
        margin-left: 0 !important;
        display: block;
        width: 100%
    }
}