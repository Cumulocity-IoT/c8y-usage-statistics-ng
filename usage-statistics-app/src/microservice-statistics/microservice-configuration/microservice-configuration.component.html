<c8y-title>{{ "Microservice Configuration" | translate }}</c8y-title>
<div [formGroup]="form" *ngIf="productCategoryList && microserviceCategoryList">
  <div class="col-md-6 col-sm-12 col-xs-12">
    <div class="card">
      <div class="card-header separator">
        <h4 class="card-title">{{ "Product Categories" | translate }}</h4>
      </div>
      <div class="card-body">
        <table class="table" style="margin: 10px; width: auto;">
          <thead>
            <tr>
              <th>{{'Product Category' | translate}}</th>
              <th>{{'CPU threshold (CPU)' | translate}}</th>
              <th>{{'Memory threshold (GiB)' | translate}}</th>
              <th>{{'Microservices' | translate}}</th>
            </tr>
          </thead>
          <tbody formArrayName="productCategoryList">
            <ng-container *ngFor="
                  let categoryForm of productCategoryList.controls;
                  let index = index
                ">
              <tr [formGroup]="categoryForm">
                <td>
                  <div *ngIf="!categoryForm.value.isEditable">
                    {{ categoryForm.value.productCategory }}
                  </div>
                  <div *ngIf="categoryForm.value.isEditable">
                    <input type="text" class="form-control" formControlName="productCategory" />
                    <c8y-messages *ngIf="
                        categoryForm.controls.productCategory.touched &&
                        categoryForm.controls.productCategory.invalid &&
                        categoryForm.controls.productCategory.errors
                      ">
                      <c8y-message>{{
                        getErrorMessage(
                        categoryForm.controls.productCategory.errors
                        ) | translate
                        }}</c8y-message>
                    </c8y-messages>
                  </div>
                </td>
                <td>
                  <ng-container *ngIf="categoryForm.value.productCategory === 'Product Services'">
                    <span class="text-muted">{{'Not Applicable' | translate}}</span>
                  </ng-container>
                  <ng-container *ngIf="categoryForm.value.productCategory !== 'Product Services'">
                    <input type="number" class="form-control" min="0" required formControlName="cpuThreshold" />
                    <c8y-messages *ngIf="
                    categoryForm.controls.cpuThreshold.touched &&
                    categoryForm.controls.cpuThreshold.invalid &&
                    categoryForm.controls.cpuThreshold.errors
                  ">
                      <c8y-message>{{
                        getErrorMessage(
                        categoryForm.controls.cpuThreshold.errors
                        ) | translate
                        }}</c8y-message>
                    </c8y-messages>
                  </ng-container>

                </td>
                <td>
                  <ng-container *ngIf="categoryForm.value.productCategory === 'Product Services'">
                    <span class="text-muted">{{'Not Applicable' | translate}}</span>
                  </ng-container>
                  <ng-container *ngIf="categoryForm.value.productCategory !== 'Product Services'">
                    <input type="number" class="form-control" min="0" required formControlName="memoryThreshold" />
                    <c8y-messages *ngIf="
                    categoryForm.controls.memoryThreshold.touched &&
                    categoryForm.controls.memoryThreshold.invalid &&
                    categoryForm.controls.memoryThreshold.errors
                  ">
                      <c8y-message>{{
                        getErrorMessage(
                        categoryForm.controls.memoryThreshold.errors
                        ) | translate
                        }}</c8y-message>
                    </c8y-messages>
                  </ng-container>
               
                </td>
                <td style="float:right">
                  <div class="prod-cat-cont-item">
                    <div>
                      <span [class]="getBadgeClass(categoryForm.value.categoryType)">{{
                        getMicroserviceCount(categoryForm.value.productCategory)
                        }}</span>
                    </div>
                    <div class="prod-cancel" [ngClass]="{
                        'hide-item': showDeleteOption(
                          categoryForm.value.categoryType
                        )
                      }">
                      <i c8yicon="minus-circle" class="m-l-4 dlt-c8y-icon-minus-circle text-danger pointer"
                        (click)="onRemoveProductCategory(index)"></i>
                    </div>
                  </div>
                </td>
              </tr>
            </ng-container>
          </tbody>
        </table>

      </div>
      <div class="card-footer separator mob">
        <button class="btn btn-default" (click)="onCancelProductCategories()">
          <i c8yicon="close" class="m-r-4 dlt-c8y-icon-close"></i> Cancel
        </button>
        <button class="btn btn-primary" (click)="onResetProductCategories()">
          <i c8yicon="reset" class="m-r-4 dlt-c8y-icon-reset"></i> Reset
        </button>
        <button class="btn btn-primary" (click)="onSaveProductCategories()" [disabled]="productCategoryList.invalid">
          <i c8yicon="save" class="m-r-4 dlt-c8y-icon-save"></i> Save
        </button>
        <button class="btn btn-primary" (click)="onAddProductCategory()"
          [disabled]="productCategoryList.value.length >= 7">
          <i c8yicon="plus-circle" class="m-r-4 dlt-c8y-icon-plus-circle"></i>
          Add Category
        </button>
      </div>

    </div>
    <div class="alert alert-info alert-info-custom" role="alert" *ngIf="productCategoryList.value.length >= 7">
      You can add a maximum of 3 custom product categories.
    </div>
  </div>
  <div class="col-md-6 col-sm-12 col-xs-12">
    <div class="card">
      <div class="card-header separator">
        <h4 class="card-title">Microservice Categorization</h4>
      </div>
      <div class="card-body micro" style="max-height: 75vh;overflow: auto;">
        <ul class="list-group" formArrayName="microserviceCategoryList">
          <ng-container *ngFor="
              let microserviceForm of microserviceCategoryList.controls;
              let index = index
            ">
            <li class="list-group-item" [formGroup]="microserviceForm">
              <div class="form-inline">
                <div class="form-group">
                  <input type="text" class="form-control" placeholder="Microservice Name"
                    formControlName="microserviceName" [size]="getInputFieldSize()" />
                  <c8y-messages *ngIf="
                      microserviceForm.controls.microserviceName.touched &&
                      microserviceForm.controls.microserviceName.invalid &&
                      microserviceForm.controls.microserviceName.errors
                    ">
                    <c8y-message>{{
                      getErrorMessage(
                      microserviceForm.controls.microserviceName.errors
                      ) | translate
                      }}</c8y-message>
                  </c8y-messages>
                </div>
                <div class="form-group">
                  <div class="c8y-select-wrapper">
                    <select class="form-control" formControlName="productCategory">
                      <ng-container *ngFor="let category of productCategoryNames">
                        <option [value]="category">{{ category }}</option>
                      </ng-container>
                    </select>
                  </div>
                </div>
                <div class="form-group">
                  <i c8yicon="minus-circle" class="m-l-4 dlt-c8y-icon-minus-circle text-danger pointer"
                    (click)="onRemoveMicroservice(index)"></i>
                </div>
              </div>
            </li>
          </ng-container>
        </ul>
      </div>

      <div class="card-footer separator mob">
        <button class="btn btn-default" (click)="onCancelMicroservice()">
          <i c8yicon="close" class="m-r-4 dlt-c8y-icon-close"></i> Cancel
        </button>
        <button class="btn btn-primary" (click)="onResetMicroservice()">
          <i c8yicon="reset" class="m-r-4 dlt-c8y-icon-reset"></i> Reset
        </button>
        <button class="btn btn-primary" (click)="onSaveMicroservice()" [disabled]="microserviceCategoryList.invalid">
          <i c8yicon="save" class="m-r-4 dlt-c8y-icon-save"></i> Save
        </button>
        <button class="btn btn-primary" (click)="onAddMicroservice()">
          <i c8yicon="plus-circle" class="m-r-4 dlt-c8y-icon-plus-circle"></i>
          Add Microservice
        </button>
      </div>
    </div>
  </div>
</div>