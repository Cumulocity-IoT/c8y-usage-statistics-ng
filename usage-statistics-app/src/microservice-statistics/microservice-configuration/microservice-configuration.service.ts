import { Injectable } from "@angular/core";
import { FormControl } from "@angular/forms";
import { FetchClient, IFetchOptions, IFetchResponse, IResult } from "@c8y/client";
import { Alert, AlertService } from "@c8y/ngx-components";

export enum CategoryType { Default = "Default", PreConfigured = "PreConfigured", Custom = "Custom" }
export enum Api { Product = "Product", Microservice = "Microservice"}
export enum PropertyName { Product = "productCategory", Microservice = "microserviceName" }
export interface ProductCategory {
    productCategory: string
    isPreconfigured: boolean
    isDefault: boolean
    categoryType: CategoryType
    isEditable: boolean,
    cpuThreshold: number,
    memoryThreshold: number
}

export interface MicroserviceCategory {
    [PropertyName.Microservice]: string
    [PropertyName.Product]: string
}

@Injectable({
    providedIn: 'root'
})
export class MicroserviceConfigurationService {

    constructor(private client: FetchClient, private alertService: AlertService) { }
    private readonly microserviceUrl: string = "/tenant/options";
    private readonly header: any = { "Content-Type": "application/json" };
    private readonly PRODUCT_KEY: string = "microservice.statistics.product.categories"
    private readonly MICROSERVICE_KEY: string = "microservice.statistics.microservice.categories"

   private async getOptions(key: Api): Promise<any> {
        try {
            let configKey: string;
            let defaultConfiguration: MicroserviceCategory[] | ProductCategory[]

            if (key === Api.Product) {
                configKey = this.PRODUCT_KEY
                defaultConfiguration = DEFAULT_PRODUCT_CATEGORIES
            }
            else if (key === Api.Microservice) {
                configKey = this.MICROSERVICE_KEY
                defaultConfiguration = DEFAULT_MICROSERVICE_CATEGORIES
            }

            const options: IFetchOptions = {
                method: "GET",
                headers: this.header,
            };
            const option = {
                category: "configuration",
                key: configKey,
            };

            const url = `${this.microserviceUrl}/${option.category}/${option.key}`;

            const res = await this.client.fetch(url, options);
            let data = await res.json();
            if (res.status !== 200) {
                await this.setOptions(key, defaultConfiguration);
                data.value = JSON.stringify(defaultConfiguration);
            }

            return JSON.parse(data.value)           
        }
        catch (error) {
            const alert: Alert = {
                text: 'Unable to get the configuration',
                type: 'danger',
                detailedData: error.message
            }
            this.alertService.add(alert)
        }

    }

    async setOptions(key: Api, configuration: ProductCategory[] | MicroserviceCategory[]): Promise<IFetchResponse> {
        try {
            let configKey: string;
            let formattedConfiguration
            if (key === Api.Product) {
                configKey = this.PRODUCT_KEY
                formattedConfiguration = configuration
            }
            else if (key === Api.Microservice) {
                configKey = this.MICROSERVICE_KEY
                formattedConfiguration = configuration
            }
            const configData = {
                category: "configuration",
                key: configKey,
                value: JSON.stringify(formattedConfiguration),
            };
            const options: IFetchOptions = {
                method: "POST",
                headers: this.header,
                body: JSON.stringify(configData),
            };
            return await this.client.fetch(`${this.microserviceUrl}`, options);
        }
        catch (error) {
            const alert: Alert = {
                text: 'Unable to update the configuration',
                type: 'danger',
                detailedData: error.message
            }
            this.alertService.add(alert)
        }

    }

    async getCategories(key: Api, isFormControlNeeded: boolean): Promise<any> {
        if(key === Api.Product){
            const categoryList = await this.getOptions(Api.Product)
            if (isFormControlNeeded) {
                return updateFormControlToUpdateOnBlur(categoryList, PropertyName.Product)
            }
            else{
                return categoryList
            }
            
        }
        else if(key === Api.Microservice){
            const categoryList = await this.getOptions(Api.Microservice)
            if (isFormControlNeeded) {
                return updateFormControlToUpdateOnBlur(categoryList, PropertyName.Microservice)
            }
            else{
                return categoryList
            }
        }     

        return null
    }

    async setDefaultMicroserviceCategories(): Promise<any> {
        await this.setOptions(Api.Microservice, DEFAULT_MICROSERVICE_CATEGORIES)
    }

    async setDefaultProductCategories(): Promise<any> {
        await this.setOptions(Api.Product, DEFAULT_PRODUCT_CATEGORIES)
    }

    getExcessProductCategories(productCategoryList: ProductCategory[]) {
        const productCategories = productCategoryList.map((elem: ProductCategory) => elem.productCategory)
        const defaultProductCategories = DEFAULT_PRODUCT_CATEGORIES.map(elem => elem.productCategory)
        return productCategories.filter(elem => !defaultProductCategories.includes(elem));
    }
}

function updateFormControlToUpdateOnBlur(list: any[], field: string): any[] {
    const listCopy = JSON.parse(JSON.stringify(list))
    listCopy.forEach(elem => {
        elem[field] = new FormControl(elem[field], { updateOn: 'blur' })
    })
    return listCopy
}

const DEFAULT_PRODUCT_CATEGORIES: ProductCategory[] = [
    {
        productCategory: 'Product Services',
        isPreconfigured: true,
        isDefault: false,
        categoryType: CategoryType.PreConfigured,
        isEditable: false,
        cpuThreshold: 0,
        memoryThreshold: 0
    },
    {
        productCategory: 'Streaming Analytics',
        isPreconfigured: true,
        isDefault: false,
        categoryType: CategoryType.PreConfigured,
        isEditable: false,
        cpuThreshold: 0,
        memoryThreshold: 0
    },
    {
        productCategory: 'Machine Learning',
        isPreconfigured: true,
        isDefault: false,
        categoryType: CategoryType.PreConfigured,
        isEditable: false,
        cpuThreshold: 0,
        memoryThreshold: 0
    },
    {
        productCategory: 'Custom Microservices',
        isPreconfigured: true,
        isDefault: true,
        categoryType: CategoryType.Default,
        isEditable: false,
        cpuThreshold: 0,
        memoryThreshold: 0
    }
]

const DEFAULT_MICROSERVICE_CATEGORIES: MicroserviceCategory[] = [
    { microserviceName: 'actility', productCategory: 'Product Services' },
    {
        microserviceName: 'administration',
        productCategory: 'Product Services'
    },
    {
        microserviceName: 'Apama EPL Apps',
        productCategory: 'Product Services'
    },
    {
        microserviceName: 'Apama Analytics Builder',
        productCategory: 'Product Services'
    },
    {
        microserviceName: 'apama-ctrl-starter',
        productCategory: 'Product Services'
    },
    {
        microserviceName: 'billwerk-agent-server',
        productCategory: 'Product Services'
    },
    { microserviceName: 'cep', productCategory: 'Product Services' },
    {
        microserviceName: 'cloud-remote-access',
        productCategory: 'Product Services'
    },
    { microserviceName: 'cockpit', productCategory: 'Product Services' },
    {
        microserviceName: 'connectivity-agent-server',
        productCategory: 'Product Services'
    },
    { microserviceName: 'datahub', productCategory: 'Product Services' },
    {
        microserviceName: 'device-simulator',
        productCategory: 'Product Services'
    },
    {
        microserviceName: 'devicemanagement',
        productCategory: 'Product Services'
    },
    {
        microserviceName: 'feature-branding',
        productCategory: 'Product Services'
    },
    {
        microserviceName: 'feature-broker',
        productCategory: 'Product Services'
    },
    {
        microserviceName: 'feature-fieldbus4',
        productCategory: 'Product Services'
    },
    {
        microserviceName: 'feature-microservice-hosting',
        productCategory: 'Product Services'
    },
    {
        microserviceName: 'feature-user-hierarchy',
        productCategory: 'Product Services'
    },
    {
        microserviceName: 'lwm2m-agent',
        productCategory: 'Product Services'
    },
    {
        microserviceName: 'opcua-mgmt-service',
        productCategory: 'Product Services'
    },
    {
        microserviceName: 'report-agent',
        productCategory: 'Product Services'
    },
    {
        microserviceName: 'smartrule',
        productCategory: 'Product Services'
    },
    {
        microserviceName: 'sms-gateway',
        productCategory: 'Product Services'
    },
    {
        microserviceName: 'sslmanagement',
        productCategory: 'Product Services'
    },
    { microserviceName: 'apama', productCategory: 'Product Services' },
    { microserviceName: 'cellid', productCategory: 'Product Services' },
    {
        microserviceName: 'datahub-webapp',
        productCategory: 'Product Services'
    },
    {
        microserviceName: 'loriot-agent',
        productCategory: 'Product Services'
    },
    {
        microserviceName: 'remote-access-agent-echo',
        productCategory: 'Product Services'
    },
    {
        microserviceName: 'retention',
        productCategory: 'Product Services'
    },
    {
        microserviceName: 'sigfox-agent',
        productCategory: 'Product Services'
    },
    {
        microserviceName: 'smartgroup',
        productCategory: 'Product Services'
    },
    {
        microserviceName: 'sslmonitor-agent-server',
        productCategory: 'Product Services'
    },
    {
        microserviceName: 'tenant-configuration',
        productCategory: 'Product Services'
    },
    {
        microserviceName: 'tenant-sla-monitoring-agent-server',
        productCategory: 'Product Services'
    },
    {
        microserviceName: 'oee-bundle',
        productCategory: 'Product Services'
    },
    {
        microserviceName: 'apama-oeeapp',
        productCategory: 'Product Services'
    },
    {
        microserviceName: 'zementis',
        productCategory: 'Machine Learning'
    },
    {
        microserviceName: 'mlw',
        productCategory: 'Machine Learning'
    },
    {
        microserviceName: 'nyoka',
        productCategory: 'Machine Learning'
    },
    {
        microserviceName: 'onnx',
        productCategory: 'Machine Learning'
    }
]
