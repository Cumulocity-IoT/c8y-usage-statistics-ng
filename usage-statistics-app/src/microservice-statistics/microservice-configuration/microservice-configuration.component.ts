import { Component, OnInit } from '@angular/core';
import { FormArray, FormBuilder, FormGroup, ValidationErrors, ValidatorFn, Validators } from '@angular/forms';
import { Alert, AlertService, ModalService, gettext } from '@c8y/ngx-components';
import { Api, CategoryType, MicroserviceCategory, MicroserviceConfigurationService, ProductCategory, PropertyName } from './microservice-configuration.service';
import $ from 'jquery'

@Component({
  selector: 'microservice-configuration',
  templateUrl: './microservice-configuration.component.html',
  styleUrls: ['./microservice-configuration.component.css']
})
export class MicroserviceConfigurationComponent implements OnInit {
  form: FormGroup;
  productCategoryNames: string[]
  constructor(
    private fb: FormBuilder,
    private modalService: ModalService,
    private microserviceConfigurationService: MicroserviceConfigurationService,
    private alertService: AlertService
  ) { }

  async ngOnInit(): Promise<void> {
    await this.loadDataInitializeForm()
    this.getInputFieldSize();
  }

  private getUniqueMicroserviceCategories(microserviceCategories: MicroserviceCategory[]): MicroserviceCategory[] {
    return microserviceCategories.filter((elem, index, self) =>
      index === self.findIndex((o) => o[PropertyName.Microservice] === elem[PropertyName.Microservice])
    );
  }

  private async loadDataInitializeForm() {
    const productCategories = [... await this.microserviceConfigurationService.getCategories(Api.Product, true)]
    const microserviceCategories = [... await this.microserviceConfigurationService.getCategories(Api.Microservice, true)]
    this.form = this.fb.group({
      productCategoryList: this.fb.array(productCategories.map(elem => this.fb.group(elem))),
      microserviceCategoryList: this.fb.array(microserviceCategories.map(elem => this.fb.group(elem)))
    })
    this.microserviceCategoryList.controls.forEach(elem => {
      elem.get(PropertyName.Microservice).addValidators(Validators.required)
      elem.get(PropertyName.Microservice).addValidators(isUniqueMicroserviceValidator(this.microserviceCategoryList))
    })

    this.productCategoryNames = this.productCategoryList.value.map((elem: ProductCategory) => elem.productCategory)
  }

  onAddProductCategory() {
    const productCategory = this.fb.group({
      productCategory: ['', [Validators.required, isUniqueProductCategoryValidator(this.productCategoryList)]],
      isPreconfigured: false,
      isDefault: false,
      isSelected: false,
      categoryType: CategoryType.Custom,
      isEditable: true,
      cpuThreshold: 0,
      memoryThreshold: 0
    })
    this.productCategoryList.push(productCategory)
  }

  async onRemoveProductCategory(index: number) {
    const excessProdCategory: ProductCategory = this.productCategoryList.at(index).value
    this.productCategoryList.removeAt(index)
    await this.onSaveProductCategories([excessProdCategory.productCategory])
  }

  async onSaveProductCategories(excessProdCategories?: string[]) {
    try {
      if (!excessProdCategories) {
        excessProdCategories = []
        const productCategoryNames = this.productCategoryList.value.map((elem: ProductCategory) => elem.productCategory)
        const microserviceCategoryListCopy = this.getUniqueMicroserviceCategories([...this.microserviceCategoryList.value]) 
        microserviceCategoryListCopy.forEach((elem: MicroserviceCategory) => {
          if (!productCategoryNames.includes(elem.productCategory)) {
            excessProdCategories.push(elem.productCategory)
          }
        })
      }
      excessProdCategories = Array.from(new Set(excessProdCategories))
      await this.microserviceConfigurationService.setOptions(Api.Product, this.productCategoryList.value)
      const microserviceCategoryList = this.getUpdatedMicroserviceCategoryList(excessProdCategories)
      await this.microserviceConfigurationService.setOptions(Api.Microservice, microserviceCategoryList)

      await this.loadDataInitializeForm()
      const alert: Alert = {
        text: 'Product categories were updated successfully',
        type: 'success',
      }
      this.alertService.add(alert)
    }
    catch (error) {
      const alert: Alert = {
        text: 'Unable to update product categories',
        type: 'danger',
        detailedData: error.message
      }
      this.alertService.add(alert)
    }
  }

  async onCancelProductCategories() {
    try {
      const productCategories = [... await this.microserviceConfigurationService.getCategories(Api.Product, true)]
      const productCategoryList = this.form.controls['productCategoryList'] as FormArray
      productCategoryList.clear()
      productCategories.map(elem => this.fb.group(elem)).forEach(elem => {
        productCategoryList.push(elem)
      });
      this.productCategoryNames = this.productCategoryList.value.map((elem: ProductCategory) => elem.productCategory)
    }
    catch (err) {
      console.log('Error ', err)
    }
  }

  async onResetProductCategories() {
    try {
      await this.modalService.confirm('Reset product category list to default state',
        `Resetting product category list to default state will remove 
                                        all configured custom categories and all microservices not falling under any of the default categories will be 
                                        moved under the product category titled Custom Microservices`,
        'warning')

      try {
        await this.microserviceConfigurationService.setDefaultProductCategories()
        const microserviceCategoryList = this.getUpdatedMicroserviceCategoryList()
        await this.microserviceConfigurationService.setOptions(Api.Microservice, microserviceCategoryList)
        await this.loadDataInitializeForm()

        const alert: Alert = {
          text: 'Product categories were reset to default successfully',
          type: 'success',
        }
        this.alertService.add(alert)
      }
      catch (error) {
        const alert: Alert = {
          text: 'Unable to reset product categories',
          type: 'danger',
          detailedData: error.message
        }
        this.alertService.add(alert)
      }
    }
    catch (error) {
      // No action
    }
  }

  onAddMicroservice() {
    const microserviceCategory = this.fb.group({
      microserviceName: ['', [Validators.required, isUniqueMicroserviceValidator(this.microserviceCategoryList)]],
      productCategory: this.productCategoryList.value.find((elem: ProductCategory) => elem.isDefault).productCategory
    })
    this.microserviceCategoryList.push(microserviceCategory)
    setTimeout(() => {
      document.getElementsByClassName('micro')[0].scrollTo(0, document.getElementsByClassName('micro')[0].scrollHeight);
    }, 100)

  }

  onRemoveMicroservice(index: number) {
    this.microserviceCategoryList.removeAt(index)
  }

  async onCancelMicroservice() {
    try {
      const microserviceCategories = [... await this.microserviceConfigurationService.getCategories(Api.Microservice, true)]
      const microserviceCategoryList = this.form.controls['microserviceCategoryList'] as FormArray
      microserviceCategoryList.clear()
      microserviceCategories.map(elem => this.fb.group(elem)).forEach(elem => {
        microserviceCategoryList.push(elem)
      });

      this.microserviceCategoryList.controls.forEach(elem => {
        elem.get(PropertyName.Microservice).addValidators(Validators.required)
        elem.get(PropertyName.Microservice).addValidators(isUniqueMicroserviceValidator(this.microserviceCategoryList))
      })
    }
    catch (err) {
      console.log('Error ', err)
    }
  }

  async onResetMicroservice() {
    try {
      await this.modalService.confirm('Reset microservice list to default state',
        `Resetting microservice list to default state will remove 
                                        all configured custom microservice categorization and the list will revert to a default configuration`,
        'warning')

      try {
        await this.microserviceConfigurationService.setDefaultMicroserviceCategories()
        await this.onCancelMicroservice()

        const alert: Alert = {
          text: 'Microservices were reset to default successfully',
          type: 'success',
        }
        this.alertService.add(alert)
      }
      catch (error) {
        const alert: Alert = {
          text: 'Unable to reset microservices',
          type: 'danger',
          detailedData: error.message
        }
        this.alertService.add(alert)
      }
    }
    catch (error) {
      // No action
    }
  }

  async onSaveMicroservice() {
    try {
      await this.microserviceConfigurationService.setOptions(Api.Microservice, this.microserviceCategoryList.value)
      await this.loadDataInitializeForm()
      const alert: Alert = {
        text: 'Microservice list was updated successfully',
        type: 'success',
      }
      this.alertService.add(alert)
    }
    catch (error) {
      const alert: Alert = {
        text: 'Unable to update microservice list',
        type: 'danger',
        detailedData: error.message
      }
      this.alertService.add(alert)
    }
  }

  getMicroserviceCount(productCategory: string): number {
    return this.microserviceCategoryList.value.filter((elem: MicroserviceCategory) => elem.productCategory === productCategory).length
  }

  getBadgeClass(categoryType: CategoryType) {
    if (categoryType === CategoryType.PreConfigured) {
      return 'badge badge-info'
    }
    else if (categoryType === CategoryType.Default) {
      return 'badge badge-system'
    }
    else if (categoryType === CategoryType.Custom) {
      return 'badge badge-success'
    }
  }

  showDeleteOption(categoryType: CategoryType) {
    return !(categoryType === CategoryType.Custom)
  }
  showEditOption(categoryType: CategoryType) {
    return categoryType === CategoryType.Custom
  }

  getErrorMessage(errors: any) {
    if (errors.required) {
      return gettext('This field is required')
    }
    else if (errors.min) {
      return gettext(`Minimum value for this field is ${errors.min.value}`);
    }
    else if (errors.isDuplicate) {
      return gettext('A field with this name already exists')
    }
    return null
  }

  get productCategoryList() {
    return this.form?.controls['productCategoryList'] as FormArray
  }

  get microserviceCategoryList() {
    return this.form?.controls['microserviceCategoryList'] as FormArray
  }

  getInputFieldSize() {
    const windowWidth = $(window).width();
    if (windowWidth > 1700) {
      return 50
    }
    else if (windowWidth > 1600) {
      return 40
    }
    else if (windowWidth > 1500) {
      return 30
    }
    return 20
  }

  private getUpdatedMicroserviceCategoryList(excessProdCategories: string[] = null): MicroserviceCategory[] {
    const microserviceCategoryListCopy = this.getUniqueMicroserviceCategories([...this.microserviceCategoryList.value])
    if (excessProdCategories && excessProdCategories.length > 0) {
      const defaultProductCategory = this.productCategoryList.value.find((elem: ProductCategory) => elem.isDefault).productCategory
      microserviceCategoryListCopy.forEach((elem: MicroserviceCategory) => {
        if (excessProdCategories.includes(elem.productCategory)) {
          elem.productCategory = defaultProductCategory
        }
      })
    }
    return microserviceCategoryListCopy
  }

}

function isUniqueProductCategoryValidator(productCategoryList: FormArray): ValidatorFn {
  return (formField: FormGroup): ValidationErrors | null => {
    const productCategoryNames = productCategoryList.value.map((elem: ProductCategory) => elem.productCategory)
    const fieldName = formField.value;
    if (productCategoryNames.includes(fieldName)) {
      return {
        isDuplicate: true
      }
    }
  };
}
function isUniqueMicroserviceValidator(microserviceCategoryList: FormArray): ValidatorFn {
  return (formField: FormGroup): ValidationErrors | null => {
    if (microserviceCategoryList.pristine) {
      return null
    }
    const microserviceCategoryNames = microserviceCategoryList.value.map((elem: MicroserviceCategory) => elem.microserviceName)
    const fieldName = formField.value;
    if (microserviceCategoryNames.includes(fieldName)) {
      return {
        isDuplicate: true
      }
    }
  };
}
