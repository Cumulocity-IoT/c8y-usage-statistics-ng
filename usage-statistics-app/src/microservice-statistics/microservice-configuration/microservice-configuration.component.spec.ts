import { NO_ERRORS_SCHEMA } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { RouterTestingModule } from '@angular/router/testing';
import { CoreModule } from '@c8y/ngx-components';
import { MicroserviceConfigurationComponent } from './microservice-configuration.component';
import { MicroserviceConfigurationService } from './microservice-configuration.service';


fdescribe('Test MicroserviceConfigurationComponent', () => {
  let component: MicroserviceConfigurationComponent;
  let fixture: ComponentFixture<MicroserviceConfigurationComponent>;
  let microserviceConfigurationService: MicroserviceConfigurationService;

  const microserviceConfigurationServiceMock = {
    setDefaultMicroserviceCategories: jest.fn(),
    setOptions: jest.fn()
}

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [RouterTestingModule, CoreModule.forRoot(), BrowserAnimationsModule],
      declarations: [MicroserviceConfigurationComponent],
      schemas: [NO_ERRORS_SCHEMA],
      providers: [
        { provide: MicroserviceConfigurationService, useValue: microserviceConfigurationServiceMock },
    ]
    });
    fixture = TestBed.createComponent(MicroserviceConfigurationComponent);
    component = fixture.componentInstance;
    microserviceConfigurationService = TestBed.inject(MicroserviceConfigurationService);
  });

  test('should be defined', () => {
    expect(fixture).toBeDefined();
  });

  test('should load expected component', () => {
    expect(fixture).toMatchSnapshot();
  });

});