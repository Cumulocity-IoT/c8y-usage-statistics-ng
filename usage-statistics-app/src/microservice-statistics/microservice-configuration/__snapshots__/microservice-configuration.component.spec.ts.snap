// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test MicroserviceConfigurationComponent should load expected component 1`] = `
<microservice-configuration
  alertService={[Function AlertService]}
  fb={[Function FormBuilder]}
  microserviceConfigurationService={[Function Object]}
  modalService={[Function ModalService]}
>
  <c8y-title>
  </c8y-title>
</microservice-configuration>
`;
